### PR TITLE
feat(#8189): Convert Add/Edit User modal to page

### PR DESCRIPTION
**Pull request by isis-meng**
# Description

Convert the Add/Edit User modal into a page with a unique URL.

medic/cht-core#8189

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* __CHT_CORE_COMPOSE_URL__
* __COUCH_SINGLE_COMPOSE_URL__
* __COUCH_CLUSTER_COMPOSE_URL__

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by isis-meng** [comment]
Some questions I have:
- What is the most appropriate way to convert error messages from modal-specific formats to be displayed for pages? (eg., lines 459-460, 474-475 in `edit-user.js`)
- What is a good way to allow information to persist in the form between refreshes?

Any advice would be really appreciated, thank you!

File: admin/src/js/controllers/edit-user.js [hunk 1]
```diff
@@ -408,22 +415,22 @@ angular
...
             $scope.editUserModel.id
           );
-          $uibModalInstance.close();
+          $state.go('users');
         })
         .catch(err => {
           if (err && err.data && err.data.error && err.data.error.translationKey) {
             $translate(err.data.error.translationKey, err.data.error.translationParams).then(function(value) {
-              $scope.setError(err, value);
+              $log.error(value, err);
```
Comment (garethbowen) [inline admin/src/js/controllers/edit-user.js:430]: `$scope.setError` used to show the error in the footer of the modal so the user could see what went wrong. It looks like this is just putting it in the console which most users won't check. Let's make sure the user gets some feedback.

File: admin/src/js/main.js [hunk 2]
```diff
@@ -246,6 +246,15 @@ angular.module('adminApp').config(function(
...
     })
+    .state('edit-user', {
+      url: '/users/edit/:name',
+      controller: 'EditUserCtrl',
+      templateUrl: 'templates/edit_user.html',
+      params: {
+        user: null,
+        name: null
+      },
+    })
```
Comment (garethbowen) [inline admin/src/js/main.js:257]: AngularJS does some magic with these state names, so if you have them as dot separated then it understands that the two pages are related. I think it'll be simpler like this...

```suggestion
    .state('user.edit', {
      url: '/edit/:name', // now that it's a child the "/users" part is included automatically.
      controller: 'EditUserCtrl',
      templateUrl: 'templates/edit_user.html',
      params: {
        user: null,
        name: null
      },
    })
```

NB: I haven't tested this at all so you may need to tweak it a bit!

File: admin/src/templates/index.html [hunk 3]
```diff
@@ -15,7 +15,7 @@
         <div class="navigation">
           <ul>
             <li ng-class="{'active-nav': checkActive('display')}" mm-auth="can_configure"><a ui-sref="display.date-time" translate>admin.display</a></li>
-            <li ui-sref-active="active-nav" mm-auth="can_configure"><a ui-sref="users" translate>Users</a></li>
+            <li ng-class="{'active-nav': checkActive('users')}" mm-auth="can_configure"><a ui-sref="users" translate>Users</a></li>
```
Comment (garethbowen) [inline admin/src/templates/index.html:18]: I don't think this change is needed if the state name is updated as per my previous suggestion.

File: admin/src/js/controllers/edit-user.js [hunk 4]
```diff
@@ -464,17 +471,20 @@ angular
       return asynchronousValidations
         .then(valid => {
           if (!valid) {
-            $scope.setError();
+            // $scope.setError();
+            $log.info("FIX ME ASYNC VALID")
```
Comment (garethbowen) [inline admin/src/js/controllers/edit-user.js:475]: If I recall correctly the `setError` function clears the "processing" state so the loading spinner goes away, the buttons are enabled again, and maybe sets some styling on the form. We'll want to include that styling here too.

**Review by garethbowen — CHANGES_REQUESTED** [review]
Thanks! This is looking great.

I've left a few tips inline.

> What is a good way to allow information to persist in the form between refreshes?

What information do you want to persist? Usually on web forms I would expect information to be lost.

**Comment by garethbowen** [comment]
@isis-meng How are you getting on here? Is there anything I can do to help?

**Comment by garethbowen** [comment]
Closing as ancient.

**Pull request closed without merging**